File: tests/wasm/test_forward_backward_exponential_oscillator.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { assertArrayClose, loadTestData } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function manualForwardBackwardExponentialOscillator(data, length, smooth) {
    const n = data.length;
    const forwardBackward = new Float64Array(n);
    const backward = new Float64Array(n);
    const histogram = new Float64Array(n);
    forwardBackward.fill(NaN);
    backward.fill(NaN);
    histogram.fill(NaN);

    const alpha = 2.0 / (smooth + 1.0);
    let ema1State = NaN;
    let ema2State = NaN;
    let prevEma2 = NaN;
    const ema1Window = [];
    const diffs = [];

    for (let i = 0; i < n; i++) {
        const value = data[i];
        if (!Number.isFinite(value)) {
            ema1State = NaN;
            ema2State = NaN;
            prevEma2 = NaN;
            ema1Window.length = 0;
            diffs.length = 0;
            continue;
        }

        ema1State = Number.isFinite(ema1State) ? alpha * value + (1.0 - alpha) * ema1State : value;
        ema1Window.push(ema1State);
        if (ema1Window.length > length) {
            ema1Window.shift();
        }

        if (ema1Window.length === length) {
            let ema2 = ema1Window[ema1Window.length - 1];
            let prev = ema2;
            let num = 0.0;
            let den = 0.0;
            for (let j = ema1Window.length - 2; j >= 0; j--) {
                ema2 += alpha * (ema1Window[j] - ema2);
                const dt = prev - ema2;
                num += dt;
                den += Math.abs(dt);
                prev = ema2;
            }
            if (den !== 0.0) {
                forwardBackward[i] = num / den * 50.0 + 50.0;
            }
        }

        ema2State = Number.isFinite(ema2State) ? alpha * ema1State + (1.0 - alpha) * ema2State : ema1State;
        if (Number.isFinite(prevEma2)) {
            diffs.push(ema2State - prevEma2);
            if (diffs.length > length) {
                diffs.shift();
            }
            if (diffs.length === length) {
                let num = 0.0;
                let den = 0.0;
                for (const diff of diffs) {
                    num += diff;
                    den += Math.abs(diff);
                }
                if (den !== 0.0) {
                    backward[i] = num / den * 50.0 + 50.0;
                    if (Number.isFinite(forwardBackward[i])) {
                        histogram[i] = (forwardBackward[i] - backward[i]) / 4.0 + 50.0;
                    }
                }
            }
        }
        prevEma2 = ema2State;
    }

    return {
        forward_backward: forwardBackward,
        backward,
        histogram,
    };
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('forward_backward_exponential_oscillator_js matches manual reference', () => {
    const close = new Float64Array(testData.close.slice(0, 180));
    const result = wasm.forward_backward_exponential_oscillator_js(close, 20, 10);
    const expected = manualForwardBackwardExponentialOscillator(close, 20, 10);

    assertArrayClose(result.forward_backward, expected.forward_backward, 1e-12, 'forward_backward mismatch');
    assertArrayClose(result.backward, expected.backward, 1e-12, 'backward mismatch');
    assertArrayClose(result.histogram, expected.histogram, 1e-12, 'histogram mismatch');
});

test('forward_backward_exponential_oscillator_batch first row matches single', () => {
    const close = new Float64Array(testData.close.slice(0, 128));
    const batch = wasm.forward_backward_exponential_oscillator_batch(close, {
        length_range: [20, 22, 2],
        smooth_range: [10, 12, 2],
    });
    const single = wasm.forward_backward_exponential_oscillator_js(close, 20, 10);

    assert.strictEqual(batch.rows, 4);
    assert.strictEqual(batch.cols, close.length);
    assert.deepStrictEqual(batch.lengths, [20, 20, 22, 22]);
    assert.deepStrictEqual(batch.smooths, [10, 12, 10, 12]);

    assertArrayClose(
        batch.forward_backward.slice(0, close.length),
        single.forward_backward,
        1e-12,
        'batch forward_backward mismatch',
    );
    assertArrayClose(
        batch.backward.slice(0, close.length),
        single.backward,
        1e-12,
        'batch backward mismatch',
    );
    assertArrayClose(
        batch.histogram.slice(0, close.length),
        single.histogram,
        1e-12,
        'batch histogram mismatch',
    );
});

test('forward_backward_exponential_oscillator_js rejects invalid params', () => {
    const close = new Float64Array(testData.close.slice(0, 64));
    assert.throws(() => {
        wasm.forward_backward_exponential_oscillator_js(close, 0, 10);
    }, /invalid length/i);
    assert.throws(() => {
        wasm.forward_backward_exponential_oscillator_js(close, 20, 0);
    }, /invalid smooth/i);
});

File: tests/python/test_forward_backward_exponential_oscillator.py
import numpy as np
import pytest

from rust_comparison import compare_with_rust
from test_utils import load_test_data

try:
    import vector_ta as ta
except ImportError:
    try:
        import vector_ta as ta
    except ImportError:
        pytest.skip(
            "Python module not built. Run 'maturin develop --features python' first",
            allow_module_level=True,
        )


def manual_forward_backward_exponential_oscillator(data, length, smooth):
    data = np.asarray(data, dtype=np.float64)
    n = data.shape[0]
    forward_backward = np.full(n, np.nan, dtype=np.float64)
    backward = np.full(n, np.nan, dtype=np.float64)
    histogram = np.full(n, np.nan, dtype=np.float64)

    alpha = 2.0 / (float(smooth) + 1.0)
    ema1_state = np.nan
    ema2_state = np.nan
    prev_ema2 = np.nan
    ema1_window = []
    diffs = []

    for i in range(n):
        value = data[i]
        if not np.isfinite(value):
            ema1_state = np.nan
            ema2_state = np.nan
            prev_ema2 = np.nan
            ema1_window.clear()
            diffs.clear()
            continue

        ema1_state = value if not np.isfinite(ema1_state) else alpha * value + (1.0 - alpha) * ema1_state
        ema1_window.append(ema1_state)
        if len(ema1_window) > length:
            ema1_window.pop(0)

        if len(ema1_window) == length:
            ema2 = ema1_window[-1]
            prev = ema2
            num = 0.0
            den = 0.0
            for hist_value in reversed(ema1_window[:-1]):
                ema2 += alpha * (hist_value - ema2)
                dt = prev - ema2
                num += dt
                den += abs(dt)
                prev = ema2
            if den != 0.0:
                forward_backward[i] = num / den * 50.0 + 50.0

        ema2_state = ema1_state if not np.isfinite(ema2_state) else alpha * ema1_state + (1.0 - alpha) * ema2_state
        if np.isfinite(prev_ema2):
            diffs.append(ema2_state - prev_ema2)
            if len(diffs) > length:
                diffs.pop(0)
            if len(diffs) == length:
                den = float(np.sum(np.abs(diffs)))
                if den != 0.0:
                    backward[i] = float(np.sum(diffs)) / den * 50.0 + 50.0
                    if np.isfinite(forward_backward[i]):
                        histogram[i] = (forward_backward[i] - backward[i]) / 4.0 + 50.0
        prev_ema2 = ema2_state

    return {
        "forward_backward": forward_backward,
        "backward": backward,
        "histogram": histogram,
    }


class TestForwardBackwardExponentialOscillator:
    @pytest.fixture(scope="class")
    def test_data(self):
        return load_test_data()

    def test_manual_reference_matches_binding(self, test_data):
        close = np.asarray(test_data["close"][:180], dtype=np.float64)
        params = {"length": 20, "smooth": 10}
        result = ta.forward_backward_exponential_oscillator(close, **params)
        expected = manual_forward_backward_exponential_oscillator(close, **params)

        np.testing.assert_allclose(
            np.asarray(result["forward_backward"], dtype=np.float64),
            expected["forward_backward"],
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(result["backward"], dtype=np.float64),
            expected["backward"],
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(result["histogram"], dtype=np.float64),
            expected["histogram"],
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_matches_rust_reference_generator(self, test_data):
        close = np.asarray(test_data["close"], dtype=np.float64)
        params = {"length": 20, "smooth": 10}
        result = ta.forward_backward_exponential_oscillator(close, **params)
        compare_with_rust(
            "forward_backward_exponential_oscillator",
            result,
            "close",
            params,
            rtol=0.0,
            atol=1e-12,
        )

    def test_stream_matches_batch(self, test_data):
        close = np.asarray(test_data["close"][:144], dtype=np.float64)
        params = {"length": 20, "smooth": 10}
        batch = ta.forward_backward_exponential_oscillator(close, **params)
        stream = ta.ForwardBackwardExponentialOscillatorStream(**params)

        out_fb = []
        out_bw = []
        out_hist = []
        for value in close:
            point = stream.update(float(value))
            if point is None:
                out_fb.append(np.nan)
                out_bw.append(np.nan)
                out_hist.append(np.nan)
            else:
                out_fb.append(point[0])
                out_bw.append(point[1])
                out_hist.append(point[2])

        np.testing.assert_allclose(
            np.asarray(out_fb, dtype=np.float64),
            np.asarray(batch["forward_backward"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(out_bw, dtype=np.float64),
            np.asarray(batch["backward"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(out_hist, dtype=np.float64),
            np.asarray(batch["histogram"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_batch_metadata_and_first_row_match_single(self, test_data):
        close = np.asarray(test_data["close"][:128], dtype=np.float64)
        result = ta.forward_backward_exponential_oscillator_batch(
            close,
            length_range=(20, 22, 2),
            smooth_range=(10, 12, 2),
        )

        assert result["rows"] == 4
        assert result["cols"] == close.shape[0]
        np.testing.assert_array_equal(result["lengths"], np.array([20, 20, 22, 22], dtype=np.uint64))
        np.testing.assert_array_equal(result["smooths"], np.array([10, 12, 10, 12], dtype=np.uint64))

        single = ta.forward_backward_exponential_oscillator(close, length=20, smooth=10)
        np.testing.assert_allclose(
            np.asarray(result["forward_backward"][0], dtype=np.float64),
            np.asarray(single["forward_backward"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(result["backward"][0], dtype=np.float64),
            np.asarray(single["backward"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )
        np.testing.assert_allclose(
            np.asarray(result["histogram"][0], dtype=np.float64),
            np.asarray(single["histogram"], dtype=np.float64),
            rtol=0.0,
            atol=1e-12,
            equal_nan=True,
        )

    def test_invalid_inputs_raise(self, test_data):
        close = np.asarray(test_data["close"][:64], dtype=np.float64)

        with pytest.raises(ValueError, match="invalid length|Invalid length"):
            ta.forward_backward_exponential_oscillator(close, length=0)
        with pytest.raises(ValueError, match="invalid smooth|Invalid smooth"):
            ta.forward_backward_exponential_oscillator(close, smooth=0)
